crates/ruff/src/jupyter/notebook.rs

```diff
@@ -130,60 +128,145 @@ impl JupyterNotebook {
+            .iter()
+            .filter(|cell| cell.cell_type == CellType::Code)
+            .count();
+
+        let mut current_offset = TextSize::from(0);
+        let mut cell_offsets = Vec::with_capacity(notebook.cells.len());
+        cell_offsets.push(TextSize::from(0));
+
+        let mut contents = Vec::with_capacity(size_hint);
+        let mut builder = JupyterIndexBuilder::default();
+
+        for (pos, cell) in notebook
+            .cells
+            .iter()
+            .enumerate()
+            .filter(|(_, cell)| cell.cell_type == CellType::Code)
+        {
+            let cell_contents = builder.add_cell(pos, cell);
+            current_offset += TextSize::of(&cell_contents) + TextSize::new(1);
+            cell_offsets.push(current_offset);
+            contents.push(cell_contents);
+        }
+
+        if cell_offsets.len() > 1 {
+            // Remove the last newline offset
+            *cell_offsets.last_mut().unwrap() -= TextSize::new(1);
+        }
+
+        Ok(Self {
+            raw: notebook,
+            index: builder.finish(),
+            content: contents.join("\n"),
+            cell_offsets,
+        })
     }
 
-    /// Concatenates all cells into a single virtual file and builds an index that maps the content
-    /// to notebook cell locations
-    pub fn index(&self) -> (String, JupyterIndex) {
-        let mut jupyter_index = JupyterIndex {
-            // Enter a line number (1-based), get back the cell (1-based)
-            // 0 index is just padding
-            row_to_cell: vec![0],
-            // Enter a line number (1-based), get back the row number in the cell (1-based)
-            // 0 index is just padding
-            row_to_row_in_cell: vec![0],
-        };
+    fn update_cell_offsets(&mut self, edits: BTreeSet<&Edit>) {
+        for edit in edits.into_iter().rev() {
+            let idx = self
+                .cell_offsets
+                .iter()
+                .tuple_windows::<(_, _)>()
+                .find_position(|(&offset1, &offset2)| {
+                    offset1 <= edit.start() && edit.end() <= offset2
+                })
+                .map_or_else(
+                    || panic!("edit outside of any cells: {edit:?}"),
+                    |(idx, _)| idx,
+                );
+
+            if edit.is_deletion() {
+                for offset in &mut self.cell_offsets[idx + 1..] {
+                    *offset -= edit.range().len();
+                }
+            } else if edit.is_insertion() {
+                let new_text_size = TextSize::of(edit.content().unwrap_or_default());
+                for offset in &mut self.cell_offsets[idx + 1..] {
+                    *offset += new_text_size;
+                }
+            } else if edit.is_replacement() {
+                let current_text_size = edit.range().len();
+                let new_text_size = TextSize::of(edit.content().unwrap_or_default());
+                match new_text_size.cmp(&current_text_size) {
```

Comment:
can't we just always add `new_text_size - current_text_size`?

Reply:
No, we can't otherwise it'll overflow as it's a `TextSize` which is internally a `u32`.

This is a replacement edit which means that the new text could be less in length as compared to the existing text.